lib/modules/manager/npm/post-update/pnpm.ts

```diff
@@ -66,16 +68,37 @@ export async function generateLockFile(
       extraEnv.NPM_AUTH = env.NPM_AUTH;
       extraEnv.NPM_EMAIL = env.NPM_EMAIL;
     }
-    const commands: string[] = [];
 
-    cmd = 'pnpm';
-    let args = 'install --recursive --lockfile-only';
+    let cmdOptions = '--recursive --lockfile-only';
     if (!GlobalConfig.get('allowScripts') || config.ignoreScripts) {
-      args += ' --ignore-scripts';
-      args += ' --ignore-pnpmfile';
+      cmdOptions += ' --ignore-scripts';
+      cmdOptions += ' --ignore-pnpmfile';
+    }
+    logger.trace({ cmdOptions }, 'pnpm command options');
+
+    if (
+      upgrades.length === 0 ||
+      !upgrades.every((upgrade) => upgrade.isLockfileUpdate)
+    ) {
+      // This command updates the lock file based on package.json
+      commands.push(`pnpm install ${cmdOptions}`);
+    }
+
+    // rangeStrategy = update-lockfile
+    const lockUpdates = upgrades.filter((upgrade) => upgrade.isLockfileUpdate);
+
+    if (lockUpdates.length) {
+      logger.debug('Performing lockfileUpdate (pnpm)');
+      // `yarn up -R` updates to the latest release in each range
+      commands.push(
+        `pnpm update --no-save ${lockUpdates
+          // TODO: types (#22198)
+          .map((update) => `${update.packageName!}@${update.newVersion!}`)
+          .filter(uniqueStrings)
+          .map(quote)
+          .join(' ')} ${cmdOptions}`,
+      );
     }
```

Comment:
this should be in else block, both commands should never be required. 🤔

Reply:
I'm not completely aware of the context in which this function is called, so I did the same as in https://github.com/renovatebot/renovate/blob/91ced24c228ad734ef3963c6b0d56536a8a11610/lib/modules/manager/npm/post-update/npm.ts#L131-L134

Can't there be an update of the lockfile due to a change in `package.json` before updating libraries ?

Reply:
if we have multiple updates in a group and one of them is a lockfile update, I think we should run `pnpm install` first and then update the lockfile

what do you think ?